include/rocksdb/cache_allocator.h

```diff
@@ -16,14 +22,35 @@ class CacheAllocator {
   // Name of the cache allocator, printed in the log
   virtual const char* Name() const = 0;
 
-  // Allocate a block of at least size size
+  // Allocate a block of at least size. Has to be thread-safe.
   virtual void* Allocate(size_t size) = 0;
-  // Deallocate previously allocated block
+
+  // Deallocate previously allocated block. Has to be thread-safe.
   virtual void Deallocate(void* p) = 0;
+
   // Returns the memory size of the block allocated at p. The default
   // implementation that just returns the original allocation_size is fine.
   virtual size_t UsableSize(void* /*p*/, size_t allocation_size) const {
     // default implementation just returns the allocation size
     return allocation_size;
   }
 };
+
+// Factory interface to obtain CacheAllocator instances.
+class CacheAllocatorFactory {
+ public:
+  virtual ~CacheAllocatorFactory() = default;
+
+  // Name of the cache allocator factory. Printed in the log.
+  virtual const char* Name() const = 0;
+
+  virtual Status NewCacheAllocator(
+      std::unique_ptr<CacheAllocator>* cache_allocator) = 0;
```

Comment:
Do we need to generate new cache allocator? Another word, do we need this factory at all? Can we just plug in an allocator to allocate blocks?

Reply:
I come up with the factory class because I want to have `ShardedCache` handle allocator sharding. In a followup patch, I'll update `ShardedCache` to shard allocator by cache key. But yeah, if we want to shard in another way (round-robin?), or don't need sharding at all, the factory class is not needed.

Reply:
What's your thoughts on the fragmentation in this per shard allocator setting?

Reply:
According to @interwq, Jemalloc by default initiate 4 * cpu_cores arenas and assign static arena id to threads. Assume we run on server with 56 cores, that's 224 arenas. Compare with we have by default 64 arenas for block cache, arguably we should have less fragmentation.

For users run multiple DB instance each with dedicated block cache, they can (and should) use a shared `CacheAllocator` for those block caches, if they wish to get the no dump option.

Reply:
Regarding the fact that with this approach, we are separating allocation for normal c++ objects and block contents. I'm arguing that normal c++ object tend to be smaller than block size. By separating allocation for them, arguably we should see fragmentation reduce (instead of increase) since mixing allocation of different size class tend to increase fragmentation.

Reply:
If each allocator class contains configurable number of arenas, it sounds great to me, it is much better than the previous idea of arena per block cache shard.

Reply:
@yiwu-arbug I'm not sure about the assumption that other allocations are all smaller than block size. RocksDB is a library, we don't know how users allocate memory. Even looking at RocksDB, large allcoation outside blocks are just as common, just to name some: malloc for reading uncompressed blocks, exponential readahead in iterators, memtable arena, compaction writeable file buffer, table builder's buffering of key hashes, indexes and bloom filter being built, etc.